### ISSUE TITLE
Issue with Unet 2D

### ISSUE THREAD

**Issue by jinxsfe** [HenriquesLab/ZeroCostDL4Mic#317]
actually, I trained the UNet 2d Successfullly in 1 years ago, but I trained again in recently and out put error is 
![image](https://github.com/HenriquesLab/ZeroCostDL4Mic/assets/121477741/b2723ea2-35fc-424d-b570-e30cca182e30)
I don't change the parameter. original data dimension is 1000

I also try to train the new dataset which dimension is 12*12, but also failed

![image](https://github.com/HenriquesLab/ZeroCostDL4Mic/assets/121477741/ca6e2140-6647-42dc-af20-2bfd00826dce)
I think no need add 3 in third dimension, the data is 32 bit unsigned integer,, the mask is 8 bit-binary. 
![image](https://github.com/HenriquesLab/ZeroCostDL4Mic/assets/121477741/91e520da-5ec2-4779-9596-e81d517c2495)

 can you give me some hint?

**Comment by IvanHCenalmor** [comment]
Dear @jinxsfe,

Not sure about the issue that you are facing. The first image is saying that it cannot find the image on the selected folder. The second one seems to be a problem with the shape of the image.

I have a new version of the notebook on a different branch (while it goes under an additional testing):

[U-Net_2D_ZeroCostDL4Mic.ipynb  ![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/HenriquesLab/ZeroCostDL4Mic/blob/bmz_model_export_update/Colab_notebooks/U-Net_2D_ZeroCostDL4Mic.ipynb)

Please try again with this notebook and if you find the same problem let us know and we will iterate over it 🤗 

Thanks again for your valuable feedback!

Iván

**Comment by jinxsfe** [comment]
Hi IvanHCnalmor:
Thanks for your a sharing for your new notebooks, yes, I had tested the UNENT2D and 3D and running successfully , here is the another question relate the notebook and my data set
my data is CryoET denoised data and original dimension is 1024*1024*151

for Unet 2D:  you had mentioned 

![Image](https://github.com/user-attachments/assets/ddb75c47-0320-436d-b852-7ae86f2c9ff5)
the #of step equivalent to the number of samples in the training set divided by the batch size, 
I calculate the number of steps below for this way, is correct? or we just need use 604/10
![Image](https://github.com/user-attachments/assets/ad6c2f21-1ee5-464f-8c80-5af3c079e986)

For Unet 3D,  I seperate original data size 1024*1024*151 into 512*512*151, add one more depth and change initial filtering from 32 to 64,  but the model still failed to catch the data's infomation, the dataset size is 24, I also set "import tensorflow as tf
from tensorflow.keras.mixed_precision import set_global_policy

# 1. Enable mixed-precision BEFORE creating your model.
set_global_policy('mixed_float16')
" to reduce load for GPU, I use two ways for trainig, first is resize to patch_size for 224*224*16, second is randomcrop patch size 64*64*64, but model still failed for catch signal even I changed depth and parameters, but I use same data for UNET2D and UNET 2D can catch siginal(original 4 layer for your note book and  5layer that I modified it) 

![Image](https://github.com/user-attachments/assets/a1a79b13-09fc-4754-9d43-c3a9b82b88dc)

**Comment by jinxsfe** [comment]
@IvanHCenalmor @paxcalpt

**Comment by IvanHCenalmor** [comment]
Hi @jinxsfe,

## Number of steps on U-Net 2D

For the U-Net 2D, for what you want, you don´t need to calculate the number of steps. As it says in the notebook: 

"**This behaviour can also be obtained by setting it to 0.**"

So setting the `number_of_steps` to 0 would be enough.

## U-Net 3D

I´m not sure what is the problem that you are facing on this notebook. Is it a memory issue? If you have a modified version of the notebook, could you please send a link to the notebook to have a better idea of which one could be the error? If you send the notebook, it would be also helpful if you keep the output of the cells to have a better insight.

Thanks a lot for the feedback,
Iván

**Comment by jinxsfe** [comment]
@IvanHCenalmor  Hi Ivan,

I had meet the two issue for UNET 2D and 3D until now.

for UNET 2D , during the unet2D lou calculation(quality), it seems it not working data type for float 32bit, 

![Image](https://github.com/user-attachments/assets/c10281e5-a0bb-4e96-a2bf-20228b5b64cd), I don't think normalized is good because it will change the pixel intensities due to CryoET is high resolution, do you have another way to solve this problems?

for UNET 3D, it's not memories issue cause the CNN can continue training, can you send me your personal email and I can send the notebook that I modified and dataset that I have.

![Image](https://github.com/user-attachments/assets/422cd1d8-694f-4154-b1fe-1ac094f76f88)validation_dice efficient is still 0.0012 so I stop training via click button

**Comment by jinxsfe** [comment]
@paxcalpt @cleterrier  @IvanHCenalmor 
it seems I change the read image method from skimage io to import tifffile, but it still show an error, the original function is "def getIoUvsThreshold(prediction_filepath, groud_truth_filepath):
  prediction = io.imread(prediction_filepath)
  ground_truth_image = img_as_ubyte(io.imread(groud_truth_filepath, as_gray=True), force_copy=True)

  threshold_list = []
  IoU_scores_list = []

  for threshold in range(0,256):
    # Convert to 8-bit for calculating the IoU
    mask = img_as_ubyte(prediction, force_copy=True)
    mask[mask > threshold] = 255
    mask[mask <= threshold] = 0

    # Intersection over Union metric
    intersection = np.logical_and(ground_truth_image, np.squeeze(mask))
    union = np.logical_or(ground_truth_image, np.squeeze(mask))
    iou_score = np.sum(intersection) / np.sum(union)

    threshold_list.append(threshold)
    IoU_scores_list.append(iou_score)

  return (threshold_list, IoU_scores_list)" i change it to 

import numpy as np
import tifffile
from skimage import img_as_ubyte

def getIoUvsThreshold(prediction_filepath, ground_truth_filepath):
    # --- 1) Read the prediction (32-bit float) with tifffile ---
    prediction_float = tifffile.imread(prediction_filepath)
    # Clamp to [0,1] in case the net produced values slightly out of range
    prediction_float = np.clip(prediction_float, 0, 1)
    # Convert to 8-bit [0..255]
    prediction_ubyte = img_as_ubyte(prediction_float)

    # --- 2) Read the ground truth (supposedly 8-bit) with tifffile ---
    ground_truth_image = tifffile.imread(ground_truth_filepath)

    # If the ground truth is truly 8-bit, ground_truth_image.dtype should be uint8.
    # But in case it's not, cast it:
    if ground_truth_image.dtype != np.uint8:
        ground_truth_image = ground_truth_image.astype(np.uint8)

    # --- 3) Loop thresholds from 0..255 and compute IoU ---
    threshold_list = []
    IoU_scores_list = []

    for threshold in range(256):
        mask = prediction_ubyte.copy()
        mask[mask > threshold] = 255
        mask[mask <= threshold] = 0

        intersection = np.logical_and(ground_truth_image, mask)
        union        = np.logical_or (ground_truth_image, mask)
        iou_score    = np.sum(intersection) / np.sum(union)

        threshold_list.append(threshold)
        IoU_scores_list.append(iou_score)

    return (threshold_list, IoU_scores_list), and result is, 

![Image](https://github.com/user-attachments/assets/e2a79cb6-5841-4e96-bccb-0ab7feffe58f)  I am still can not see the input image, the image should be like 

![Image](https://github.com/user-attachments/assets/cf742a90-482f-4739-ab6b-9cce03ec765b) which is 32 bit float.

**Comment by jinxsfe** [comment]
this is training progress  

![Image](https://github.com/user-attachments/assets/d6af4b8a-254c-4410-a217-4d8bcdc7acb8)

**Comment by jinxsfe** [comment]
https://colab.research.google.com/drive/1BK30TzmHLH3tUCSbFlgxNQ6KtcT0AG-o?usp=drive_link, https://colab.research.google.com/drive/1z23SRj427vz7tpP8szZp7noiqZt8CdL8?usp=drive_link

**Comment by esgomezm** [comment]
Hi @jinxsfe it seems that there are several issues around the same problem. I'm closing this issue to avoid redundancy
